pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -54,166 +56,193 @@ func NewDockercfgController(cl client.Interface, options DockercfgControllerOpti
+				serviceAccount := obj.(*api.ServiceAccount)
+				glog.V(4).Infof("Adding service account %s", serviceAccount.Name)
+				e.enqueueServiceAccount(serviceAccount)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				serviceAccount := cur.(*api.ServiceAccount)
+				glog.V(4).Infof("Updating service account %s", serviceAccount.Name)
+				// Resync on service object relist.
+				e.enqueueServiceAccount(serviceAccount)
+			},
 		},
 	)
 
+	e.syncHandler = e.syncServiceAccount
 	e.dockerURL = options.DefaultDockerURL
 
 	return e
 }
 
 // DockercfgController manages dockercfg secrets for ServiceAccount objects
 type DockercfgController struct {
-	stopChan chan struct{}
-
 	client client.Interface
 
 	dockerURL     string
 	dockerURLLock sync.Mutex
 
+	serviceAccountCache      cache.Store
 	serviceAccountController *framework.Controller
+
+	queue workqueue.RateLimitingInterface
+
+	// syncHandler does the work. It's factored out for unit testing
+	syncHandler func(serviceKey string) error
+}
+
+func (e *DockercfgController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	go e.serviceAccountController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go wait.Until(e.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Infof("Shutting down dockercfg secret controller")
+	e.queue.ShutDown()
 }
 
-// Runs controller loops and returns immediately
-func (e *DockercfgController) Run() {
-	if e.stopChan == nil {
-		e.stopChan = make(chan struct{})
-		go e.serviceAccountController.Run(e.stopChan)
+func (e *DockercfgController) enqueueServiceAccount(serviceAccount *api.ServiceAccount) {
+	if !needsDockercfgSecret(serviceAccount) {
+		return
 	}
+
+	key, err := controller.KeyFunc(serviceAccount)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", serviceAccount, err)
+		return
+	}
+
+	e.queue.Add(key)
 }
 
-// Stop gracefully shuts down this controller
-func (e *DockercfgController) Stop() {
-	if e.stopChan != nil {
-		close(e.stopChan)
-		e.stopChan = nil
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (e *DockercfgController) worker() {
+	for {
+		if !e.worker_inner() {
+			return
+		}
 	}
 }
 
+// worker_inner returns true if the worker thread should continue
+func (e *DockercfgController) worker_inner() bool {
+	key, quit := e.queue.Get()
+	if quit {
+		return false
+	}
+	defer e.queue.Done(key)
+
+	if err := e.syncHandler(key.(string)); err == nil {
+		// this means the request was successfully handled.  We should "forget" the item so that any retry
+		// later on is reset
+		e.queue.Forget(key)
+
+	} else {
+		// if we had an error it means that we didn't handle it, which means that we want to requeue the work
+		utilruntime.HandleError(fmt.Errorf("error syncing service, it will be retried: %v", err))
+		e.queue.AddRateLimited(key)
```

Comment:
> max number of retries?

Unless we flag permanent failures, we'll just pick up on the next resync.  I wasn't prepared to say that.